### DISCUSSION TITLE
Help me to understand Next.js with + redux + saga / thunk

### DISCUSSION THREAD

**Discussion by TheBigMoon** [vercel/next.js#14401] [Help] [answered]
I can`t understand how can i apply redux-saga or redux-thunk with Next.js
The examples in next.js developers git ([saga ](https://github.com/vercel/next.js/tree/canary/examples/with-redux-saga)and [thunk](https://github.com/vercel/next.js/tree/canary/examples/with-redux-thunk)) is so hard to me fro understanding.
I tried so much different solutions from many sites and examples.
May be someone can help me and show example here or your git.
I want to see realy simple saga or thunk examples
May be you have git repo or can show me here.
P.S.: sorry for my english if something is not correct

**Comment by mattcarlotta** [comment, marked as answer]
There are two main questions you should ask yourself before jumping into Redux:
1.) Why do I need Redux? Can I accomplish the same thing with React state?
2.) Is what I need to accomplish going to be asynchronous or synchronous?

If you're unable to answer any of the questions above, then you most likely don't need to use Redux and Redux Thunk/Saga in your application. Too often novice developers will jump into Redux without understanding why they need it and the obstacles/complexities it introduces when implemented within your app.

That said, you'll need to be more specific as to what you're trying to accomplish in your app and what you don't understand about the example implementation. The examples provided aren't really meant to be tutorials, but guides on how to set up your project with Redux. Without a reference point of what you're trying to accomplish, it's going to be hard to provide relevant guidance.

---

With the above out of the way, let's take a look at the [with-redux-saga](https://github.com/vercel/next.js/tree/canary/examples/with-redux-saga) example and how it utilizes [next-redux-wrapper](https://github.com/kirill-konshin/next-redux-wrapper) to set up Redux. 

In order to set up Redux, the `next-redux-wrapper` needs to wrap the [_app page](https://github.com/vercel/next.js/blob/canary/examples/with-redux-saga/pages/_app.js#L7) with a redux `Provider`. The package obscures this by providing a [withRedux](https://github.com/kirill-konshin/next-redux-wrapper/blob/master/packages/wrapper/src/index.tsx#L151-L218) method that internally wraps the `App` function with a `Provider` and a `store`:

```js
return (
  <Provider store={store.current}>
    <Component {...initialProps} {...resultProps} />
  </Provider>
);
```

However, before we can do the above, we need to set up a saga implementation by creating a [makeStore](https://github.com/vercel/next.js/blob/canary/examples/with-redux-saga/store.js#L16-L23) function. Then, we'll utilize this `makeStore` function by creating a reusable [wrapper](https://github.com/vercel/next.js/blob/canary/examples/with-redux-saga/store.js#L25) object. This `wrapper` will have several methods attached to it `wrapper.withRedux`, `wrapper.getStaticPaths`, `wrapper.getStaticProps`, and `wrapper.getServerSideProps`. The `withRedux` wrapper will wrap the `_app` page (as mentioned above). The other methods will allow us to access `store` (`context.store`) by providing methods that act the same as Next's [data fetching](https://nextjs.org/docs/basic-features/data-fetching) methods.

Now that our app is set up with Redux and Redux Saga, we'll import this reusable [wrapper](https://github.com/vercel/next.js/blob/canary/examples/with-redux-saga/pages/index.js#L4) object within a page, then we'll utilize one of the methods mentioned above. In this case, we're using [wrapper.getStaticProps](https://github.com/vercel/next.js/blob/canary/examples/with-redux-saga/pages/index.js#L18-L27), which appends the `store` property to the `context` object (`context.store` is [destructured](https://hacks.mozilla.org/2015/05/es6-in-depth-destructuring/) to `({ store })`).  

Now that we have the Redux `store` property available to us, we can `dispatch` actions like [tickClock](https://github.com/vercel/next.js/blob/ee1cf6f8c4a09d1660ffe7655d86cc604e21008d/examples/with-redux-saga/actions.js#L47-L53) which will determine whether or not this action has been run from the client or from the server. Additionally, if the store doesn't have `placeholderData` state, then we can `dispatch` [loadData](https://github.com/vercel/next.js/blob/ee1cf6f8c4a09d1660ffe7655d86cc604e21008d/examples/with-redux-saga/actions.js#L32-L34) to request some placeholder user data from an API (after dispatching this `action`, the [loadDataSaga](https://github.com/vercel/next.js/blob/ee1cf6f8c4a09d1660ffe7655d86cc604e21008d/examples/with-redux-saga/saga.js#L12-L20) runs). The tasks are essentially wrapped in a [Promise](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise) to render block the `Index` component from mounting until all actions have been resolved.

Once the sagas have resolved, the [Index](https://github.com/vercel/next.js/blob/canary/examples/with-redux-saga/pages/index.js#L8-L16) component is mounted and then the clock is started by dispatching the [startClock](https://github.com/vercel/next.js/blob/ee1cf6f8c4a09d1660ffe7655d86cc604e21008d/examples/with-redux-saga/actions.js#L43-L45) action, which runs the [runClockSaga](https://github.com/vercel/next.js/blob/ee1cf6f8c4a09d1660ffe7655d86cc604e21008d/examples/with-redux-saga/saga.js#L4-L10). Now when the [Page](https://github.com/vercel/next.js/blob/canary/examples/with-redux-saga/components/page.js) component is mounted within `Index`, it'll have access to this data (a running clock and some placeholder user data).

What's the point of the above example? To persist a running clock and to persist asynchronous fetched data across pages. If you want to access this persisted data, then you'll either need to use `react-redux` to `connect` to the component, or use the `useSelector` hook.

**Reply by TheBigMoon** [reply]
I think i got it.
I`ll write after my try
P.S.: Your description is the best I've ever read. Thank you so much.

**Reply by TheBigMoon** [reply]
Something going wrong
I think i make all just like you said and with links-examples
I have white screen and nothing else
My IDE console showing "compiled successfully"
May be u can look at my repo with this problem?
[Repo](https://github.com/TheBigMoon/nextJS-posts-app)
You can push your modifications

**Reply by mattcarlotta** [reply]
You're close, you're just missing the redux-saga `END` action after your API request has been fulfilled. More info about END can be found [here](https://redux-saga.js.org/docs/advanced/Channels.html) -- in short, `toPromise` opens a blocking channel and `END` closes the blocking channel.

You can `dispatch` END [here](https://github.com/TheBigMoon/nextJS-posts-app/blob/master/pages/index.tsx#L27-L31):
```js
import { END } from "redux-saga";

...

export const getServerSideProps = wrapper.getServerSideProps(
  async ({ store }) => {
    store.dispatch(getPosts());
    store.dispatch(END); // dispatch it here
    // @ts-ignore
    await store.sagaTask.toPromise();
  }
);
```

**OR**

`yield` END [here](https://github.com/TheBigMoon/nextJS-posts-app/blob/master/sagas/sagas.ts#L23-L26):
```js
import { END } from "redux-saga";

...

function* getPostsWorker() {
  const posts = yield call(API.getPosts);
  yield put(setPosts(posts));
  yield put(END); // yield it here
}
```

**Reply by TheBigMoon** [reply]
Wow! The error is gone!
But can u tell me: if all my dispatchs in getServerSideProps is SSR, why i cant see the posts?
If asking in another way: why i`m geting posts = null from initialStore instead of posts from request. I thougth i make all good

**Reply by mattcarlotta** [reply]
You'll also need to `import { HYDRATE } from 'next-redux-wrapper'` and add the `HYDRATE` action within the reducer. See [example](https://github.com/vercel/next.js/blob/canary/examples/with-redux-saga/reducer.js#L14-L16) for implementation. If you have multiple reducers, then it'll be `action.payload.reducerState` where `reducerState` refers to the property that a specific reducer state resides in.

For example, if you're using multiple reducers:
```combineReducers({ posts: postReducer, users: usersReducer });```

Within the `postReducer` the `HYDRATE` case will be: 
```js 
return { ...state, ...action.payload.posts };
``` 
Within the `usersReducer` the `HYDRATE` case will be:
 ```js
return { ...state, action.payload.users };
```
and so on...

**Reply by TheBigMoon** [reply]
OMG! ITS ALIVE
I cant tell you how much u made for me
I`m spend 10+ hours to understand it and u just tell about all
I just say A BIG thank you
You are now my hero

**Comment by sohebalam** [comment]
How do you do this with thunk?
